### PR TITLE
[core][Android] Fix activity contract registration after host destruction

### LINKED ISSUE #19512: Expo ImagePicker fails to launch. Call to function 'ExponentImagePicker.launchCameraAsync' has been rejected.
### Summary

My users on Android arrive at a point where they can no longer launch the camera.
User has given camera and media library permissions.
User was able to open camera.
User is no longer able to open Camera or Media Library.

It seems to be related to the expo task manager too.

** I've found that the issue appears when: ** 
- Given you are tracking location via a task in the task manager
- When you close the app
- When you reopen the app
- When you try to pick an image
- Then the image picker will not open

I've found that the issue does **not** appear when location tracking is not active when the app was last closed. 

Check the linked repo with reproducible example.

``` 
result = await ImagePicker.launchCameraAsync({
        base64: true,
        allowsEditing: true,
        mediaTypes: ImagePicker.MediaTypeOptions.Images,
        quality: 0.2
      });

// or

result = await ImagePicker.launchImageLibraryAsync({...})
```

Resulting error logs (source Sentry):
```
Error: Error: Call to function 'ExponentImagePicker.launchCameraAsync' has been rejected.
→ Caused by: java.lang.IllegalStateException: Attempting to launch an unregistered ActivityResultLauncher with contract expo.modules.imagepicker.contracts.CameraContra...
  at ? (app:///index.android.bundle:3411:1320)
  at generatorResume([native code])
  at n(app:///index.android.bundle:269:69)
  at p(app:///index.android.bundle:269:316)
  at u(app:///index.android.bundle:2634:157)
  at ? (app:///index.android.bundle:2634:884)
  at ? (app:///index.android.bundle:76:1692)
  at p(app:///index.android.bundle:76:528)
  at N(app:///index.android.bundle:76:918)
  at callReactNativeMicrotasks(app:///index.android.bundle:76:3079)
  at value(app:///index.android.bundle:56:2935)
  at ? (app:///index.android.bundle:56:1045)
  at value(app:///index.android.bundle:56:2583)
  at value(app:///index.android.bundle:56:1004)
  at value([native code])
  at value([native code])
```

Closing and reopening the app does not resolve the issue.
Once a user has met this error, only reinstalling the app works.

### What platform(s) does this occur on?

Android

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 12.0.1
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 16.12.0 - ~/.nvm/versions/node/v16.12.0/bin/node
      Yarn: 1.22.17 - /usr/local/bin/yarn
      npm: 8.1.0 - ~/.nvm/versions/node/v16.12.0/bin/npm
      Watchman: 2021.06.07.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.10.1 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    IDEs:
      Android Studio: 4.1 AI-201.8743.12.41.7199119
      Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^46.0.0 => 46.0.9 
      react: 18.0.0 => 18.0.0 
      react-dom: 18.0.0 => 18.0.0 
      react-native: 0.69.6 => 0.69.6 
      react-native-web: ~0.18.7 => 0.18.8 
      react-navigation: ^4.4.4 => 4.4.4 
    npmGlobalPackages:
      eas-cli: 0.53.1
      expo-cli: 6.0.6
    Expo Workflow: managed

### Minimal reproducible example

The error appears only for production builds and dev builds using a dev-client.  
I've been able to reproduce it via the example demo app in this github repository: https://github.com/Nxtra/expo-image-picker-issue-reproduce  

The steps to reproduce the exact issue are in the repo, also here:
- run `npx expo start`
- open via dev client
- Press "pick an image"
- You will be able to pick an image
- Select start tracking
- After a few sec an icon will appear in you top notification bar indicating that you are being tracked
- Close the app (really close it by swiping it out of your open apps menu)
- Open app
- Press "pick an image"
- **The camera picker won't open!**

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Fixed https://github.com/expo/expo/issues/19512.

# How

When the host is destroyed and the app isn't killed, contracts are not registered with the new Activity instance after the system recreates the main Activity. I've changed the order of initialization, and now, it should re-register all contracts correctly. 

# Test Plan

- NCL (expo-image-picker example) with `don't keep activity` option ✅

### CODE REVIEW

**Review by alanjhughes — APPROVED** [review]

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt [hunk 1]
```diff
@@ -63,6 +63,8 @@ class AppContext(
   val registry = ModuleRegistry(WeakReference(this))
   private val reactLifecycleDelegate = ReactLifecycleDelegate(this)
 
+  private var hostWereDestroyed = false
```
Comment (fobos531) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt:66]:
```suggestion
  private var hostWasDestroyed = false
```

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt [hunk 2]
```diff
@@ -329,6 +332,12 @@ class AppContext(
       "Current Activity is of incorrect class, expected AppCompatActivity, received ${currentActivity?.localClassName}"
     }
 
+    // We need to re-register activity contracts when reusing AppContext with new Activity after host destruction.
+    if (hostWereDestroyed) {
+      hostWereDestroyed = false
```
Comment (fobos531) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt:337]:
```suggestion
      hostWasDestroyed = false
```

File: packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt [hunk 3]
```diff
@@ -329,6 +332,12 @@ class AppContext(
       "Current Activity is of incorrect class, expected AppCompatActivity, received ${currentActivity?.localClassName}"
     }
 
+    // We need to re-register activity contracts when reusing AppContext with new Activity after host destruction.
+    if (hostWereDestroyed) {
```
Comment (fobos531) [inline packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt:336]:
```suggestion
    if (hostWasDestroyed) {
```

**Comment by jparkrr** [comment]
Am I understanding [the changelog](https://github.com/expo/expo/blob/main/packages/expo-modules-core/CHANGELOG.md) correctly, that this hasn't been published yet?

**Comment by lukmccall** [comment]
That fix was published as `expo@50.0.12`

**Pull request merged**